core/src/main/java/hudson/SystemProperties.java

```diff
@@ -0,0 +1,220 @@
+ * 
+ * @author Johannes Ernst
+ */
+public class SystemProperties {
+    /**
+     * The ServletContext to get the init parameters from.
+     */
+    private static ServletContext theContext;
+
+    /**
+     * Logger.
+     */
+    private static final Logger LOGGER = Logger.getLogger(SystemProperties.class.getName());
+
+    /**
+     * Gets the system property indicated by the specified key.
+     * This behaves just like System.getProperty(String), except that it
+     * also consults the ServletContext's init parameters.
+     * 
+     * @param      key   the name of the system property.
+     * @return     the string value of the system property,
+     *             or <code>null</code> if there is no property with that key.
+     *
+     * @exception  SecurityException  if a security manager exists and its
+     *             <code>checkPropertyAccess</code> method doesn't allow
+     *              access to the specified system property.
+     * @exception  NullPointerException if <code>key</code> is
+     *             <code>null</code>.
+     * @exception  IllegalArgumentException if <code>key</code> is empty.
+     */
+    public static String getProperty(String key) {
+        String value = System.getProperty(key); // keep passing on any exceptions
+        if( value != null ) {
+            if( LOGGER.isLoggable(Level.INFO )) {
+                LOGGER.log(Level.INFO, "Property (system): {0} => {1}", new Object[]{ key, value });
+            }
+        } else if( theContext != null ) {
+            value = theContext.getInitParameter(key);
+            if( value != null ) {
+                if( LOGGER.isLoggable(Level.INFO )) {
+                    LOGGER.log(Level.INFO, "Property (context): {0} => {1}", new Object[]{ key, value });
+                }
+            }
+        } else {
+            if( LOGGER.isLoggable(Level.INFO )) {
+                LOGGER.log(Level.INFO, "Property (not found): {0} => {1}", new Object[]{ key, value });
+            }
+        }
+
+        return value;
+    }
+
+    /**
+     * Gets the system property indicated by the specified key.
+     * This behaves just like System.getProperty(String), except that it
+     * also consults the ServletContext's init parameters.
+     * 
+     * @param      key   the name of the system property.
+     * @param      def   a default value.
+     * @return     the string value of the system property,
+     *             or <code>null</code> if there is no property with that key.
+     *
+     * @exception  SecurityException  if a security manager exists and its
+     *             <code>checkPropertyAccess</code> method doesn't allow
+     *              access to the specified system property.
+     * @exception  NullPointerException if <code>key</code> is
+     *             <code>null</code>.
+     * @exception  IllegalArgumentException if <code>key</code> is empty.
+     */
+    public static String getProperty(String key, String def) {
```

Comment:
Would make more sense as `getString(…)` or just `get(…)` given the class name and names of other methods.